cpp/src/arrow/array.h

```diff
@@ -229,6 +228,32 @@ ARROW_EXPORT
 Status MakeArrayOfNull(const std::shared_ptr<DataType>& type, int64_t length,
                        std::shared_ptr<Array>* out);
 
+/// \brief Create a strongly-typed Array instance with all elements null
+/// \param[in] pool the pool from which memory for this array will be allocated
+/// \param[in] type the array type
+/// \param[in] length the array length
+/// \param[out] out resulting Array instance
+ARROW_EXPORT
+Status MakeArrayOfNull(MemoryPool* pool, const std::shared_ptr<DataType>& type,
+                       int64_t length, std::shared_ptr<Array>* out);
+
+/// \brief Create an Array instance whose slots are the given scalar
+/// \param[in] scalar the value with which to fill the array
+/// \param[in] length the array length
+/// \param[out] out resulting Array instance
+ARROW_EXPORT
+Status MakeArrayFromScalar(const Scalar& scalar, int64_t length,
+                           std::shared_ptr<Array>* out);
+
+/// \brief Create a strongly-typed Array instance with all elements null
+/// \param[in] pool the pool from which memory for this array will be allocated
+/// \param[in] scalar the value with which to fill the array
+/// \param[in] length the array length
+/// \param[out] out resulting Array instance
+ARROW_EXPORT
+Status MakeArrayFromScalar(MemoryPool* pool, const Scalar& scalar, int64_t length,
+                           std::shared_ptr<Array>* out);
```

Comment:
Seems like you might want to put some of these utilities into `src/arrow/array/` instead of cramming more stuff into `array.cc`

Reply:
(does not need to be fixed in this patch)

Reply:
See https://issues.apache.org/jira/browse/ARROW-6493